File: dice_to_n/dice_to_n.cpp
// dice_to_n.cpp : This file contains the 'main' function. Program execution begins and ends there.
// Írj egy numerikus szimulációt, ami a következő játékot játsza. Az 0-s mezőből indulsz és annyit lépsz előre, ahanyast dobsz egy dobókockával. A kérdés, hogy az összes játékhoz képest a játékok mekkora részében lépsz egy előre megadott mezőbe? Legyen ennek az értéke N! Ennek segítségével megadható, hogy empirikusan mekkora a valószínűsége, hogy egy adott mezőbe lépsz. A program vegye az első és egyetlen hívási argumentumot az N értékének, vagy ha nincs megadva hívási paraméterként, akkor kérje be a standard inputról. Majd írja ki, hogy 10, 100, ... stb játékra vetítve az esetek hanyad részében lép a program N-be!

#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <random>

bool isPowerOfTen(unsigned long int input) {
    return
        input == 1L
        || input == 10L
        || input == 100L
        || input == 1000L
        || input == 10000L
        || input == 100000L
        || input == 1000000L
        || input == 10000000L
        || input == 100000000L
        || input == 1000000000L
        || input == 10000000000L
        || input == 100000000000L
        || input == 1000000000000L;
}

int main(int argc, char** argv)
{
    unsigned int N;
    if (argc > 1)
        N = atoi(argv[1]);
    else
        std::cin >> N;

    std::mt19937_64 engine(1000);
    std::uniform_int_distribution<int> distr(1, 6);

    long int stepIn = 0;                                   // how many times did I step into the desired N

    for (unsigned long int i = 0; i < 4294967294; ++i)     // i: number of games
    {
        unsigned long state = 0;                           // state of the game
        while (state < N)
            state += distr(engine);

        if (state == N)
            ++stepIn;

        if (isPowerOfTen(i))
        {
            int prec = int(log(sqrt(i)) / log(10)) + 3;
            std::cout.precision(prec);
            std::cout << stepIn / double(i) << " +- " << 1 / sqrt(i) << std::endl;

        }
    }

    return 0;
}

File: py_interview/dice_to_n.py
"""Írj egy numerikus szimulációt, ami a következő játékot játsza.

A 0-s mezőből indulsz és annyit lépsz előre, ahanyast dobsz egy dobókockával.
A kérdés, hogy az összes játékhoz képest a játékok mekkora részében lépsz
egy előre megadott mezőbe? Legyen ennek az értéke N!
Ennek segítségével megadható, hogy empirikusan mekkora a valószínűsége,
hogy egy adott mezőbe lépsz.

A program vegye az első hívási argumentumot az N értékének,
vagy ha nincs megadva hívási paraméterként, akkor kérje be a standard inputról.
Majd írja ki, hogy 10, 100, ... stb játékra vetítve, hogy
az esetek hanyad részében lép a program N-be!

Lehet megadni második paramétert is, amely megadja, hogy teljes
eloszlást számoljon-e és ábrázoljon a program N-ig.
Az 1, y és i értékek igazat jelentenek, minden más hamisat."""

import random
import sys
from collections import defaultdict
import matplotlib.pyplot as plt
random.seed(0)


def hit(target: int) -> int:
    """If dice hits, return 1, 0 otherwise."""
    dice_sum = 0
    while dice_sum < target:
        dice_sum += random.randint(1, 6)
        if dice_sum == target:
            return 1
    return 0


def single_hit(target: int) -> None:
    total_hit = 0

    iterations = 0
    itermax = 10
    while True:
        for _ in range(itermax):
            total_hit += hit(target)
            iterations += 1
        itermax *= 2
        print(total_hit/iterations)


def distr(target: int) -> None:
    hits = defaultdict(lambda: 0)
    itermax = 10
    while True:
        for _ in range(itermax):
            total = 0
            while total <= target:
                hits[total] += 1
                total += random.randint(1, 6)
        itermax *= 2
        fig, ax = plt.subplots()
        ax.bar(list(hits.keys())[1:], [val/hits[0]
               for val in hits.values()][1:], width=1.)
        ax.axhline(2/7, linestyle='-', color='k')
        ax.set_xticks(range(1, target, target//6))
        plt.savefig("plt.png")
        input("press any key")


def main() -> None:
    if len(sys.argv) > 1:
        target = int(sys.argv[1])
    else:
        target = int(input("Target value: "))
    if len(sys.argv) > 2:
        distr_str = sys.argv[2]
    else:
        distr_str = input("Do you want distribution plot (y/n)? ")

    if distr_str in ["y", "1", "i"]:
        distr(target)

    single_hit(target)


if __name__ == "__main__":
    sys.exit(main())
